x/superfluid/keeper/stake_test.go

```diff
@@ -1028,6 +1031,677 @@ func (s *KeeperTestSuite) TestRefreshIntermediaryDelegationAmounts() {
+
+			// Lock check
+			s.lockCheck(s.Ctx, *lock, valAddr.String(), true)
+
+			// Bank check
+			balanceAfterConvertLockToStake := s.App.BankKeeper.GetAllBalances(s.Ctx, sender)
+			s.Require().True(balanceBeforeConvertLockToStake.IsEqual(balanceAfterConvertLockToStake))
+		})
+	}
+}
+
+func (s *KeeperTestSuite) TestConvertUnlockedToStake() {
+	defaultJoinTime := s.Ctx.BlockTime()
+	type tc struct {
+		usePartialShares    bool
+		useMinAmountToStake bool
+		useNonGammPrefix    bool
+		expectedError       error
+	}
+	testCases := map[string]tc{
+		"convert unlocked gamm shares": {},
+		"convert partial shares": {
+			usePartialShares: true,
+		},
+		"min amount to stake exceeds exit pool amount": {
+			useMinAmountToStake: true,
+			expectedError: types.TokenConvertedLessThenDesiredStakeError{
+				ActualTotalAmtToStake:   sdk.NewInt(8309),
+				ExpectedTotalAmtToStake: sdk.NewInt(999999999),
+			},
+		},
+		"error: use non gamm prefix": {
+			useNonGammPrefix: true,
+			expectedError: types.TokenConvertedLessThenDesiredStakeError{
+				ActualTotalAmtToStake:   sdk.NewInt(8309),
+				ExpectedTotalAmtToStake: sdk.NewInt(999999999),
+			},
+		},
+	}
+
+	for name, tc := range testCases {
+		s.Run(name, func() {
+			s.SetupTest()
+			s.Ctx = s.Ctx.WithBlockTime(defaultJoinTime)
+
+			// We bundle all migration setup into a single function to avoid repeating the same code for each test case.
+			_, _, _, _, sender, poolId, shareOut, _ := s.SetupUnbondConvertAndStakeTest(s.Ctx, false, false, false, true)
+
+			// testing params
+			valAddr := s.SetupValidator(stakingtypes.Bonded)
+			minAmtToStake := sdk.ZeroInt()
+			if tc.useMinAmountToStake {
+				minAmtToStake = sdk.NewInt(9999999999)
+			}
+			sharesToStake := shareOut
+			if tc.usePartialShares {
+				sharesToStake.Amount = sharesToStake.Amount.Quo(sdk.NewInt(2))
+			}
+			if tc.useNonGammPrefix {
+				sharesToStake = sdk.NewInt64Coin("foo", 10)
+			}
+
+			balanceBeforeConvert := s.App.BankKeeper.GetBalance(s.Ctx, sender, shareOut.Denom)
+			s.Require().True(!balanceBeforeConvert.Amount.IsZero())
+
+			bondDenom := s.App.StakingKeeper.BondDenom(s.Ctx)
+			totalPoolLiquidityBeforeConvert, err := s.App.GAMMKeeper.GetTotalPoolLiquidity(s.Ctx, poolId)
+			s.Require().NoError(err)
+			bondDenomPoolAmtBeforeConvert := totalPoolLiquidityBeforeConvert.AmountOf(bondDenom)
+
+			// system under test
+			totalAmtConverted, err := s.App.SuperfluidKeeper.ConvertUnlockedToStake(s.Ctx, sender, valAddr.String(), sharesToStake, minAmtToStake)
+			if tc.expectedError != nil {
+				s.Require().Error(err)
+				return
+			}
+			s.Require().NoError(err)
+
+			// gamm check
+			totalPoolLiquidityAfterConvert, err := s.App.GAMMKeeper.GetTotalPoolLiquidity(s.Ctx, poolId)
+			s.Require().NoError(err)
+			// check that pool liquidity have reduced
+			bondDenomPoolAmtAfterConvert := totalPoolLiquidityAfterConvert.AmountOf(bondDenom)
+			s.Require().True(bondDenomPoolAmtAfterConvert.LT(bondDenomPoolAmtBeforeConvert))
```

Comment:
We should know exactly how much this is reduced by right? Would be nice to check the exact value rather than inequalities if possible.

Reply:
Added in https://github.com/osmosis-labs/osmosis/pull/5949/commits/fe4b64c8b1e1d22c45d3492539dfeb244ec1fcb0